SLING-8266 - Stop embedding ESAPI

rombert opened the pull request.

rombert committed:
SLING-8266 - Stop embedding ESAPI

Delegate XSS handling to the Sling XSS API instead of using ESAPI
directly.

Tests pending.

rombert committed:
SLING-8266 - Stop embedding ESAPI

Fix formatting.

rombert committed:
SLING-8266 - Stop embedding ESAPI

Remove unused import

rombert committed:
SLING-8266 - Stop embedding ESAPI

Corect version of API bundle dependency.

rombert commented:
This sketches how I think it should work (not tested yet!) but would appreciate the review.

CI check will fail since tests are not updated (working on it).

kwin commented:
Is there a commit missing here? XssSupport used to have only static methods, but now was turned into an OSGi component. All tags need to get and unget a reference to the service... This is still missing.

rombert commented:
@kwin - it is an OSGi component only to obtain a reference to the `XSSAPI` field. The static method is still there, but deprecated. So everything will still work as before.

But your comment made me think and probably the (new) instance method no longer makes sense, and I should just drop it.

kwin commented:
@rombert I don't think this will work if the component is not "immediate". If noone is requesting a reference to `XssSupport` that component will never be started! I would recommend rather getting the service lazily in the tag and unget it in `release()` (https://docs.oracle.com/javaee/6/api/javax/servlet/jsp/tagext/Tag.html#release()).

klcodanr commented:
@kwin would that cause any issues with the usage from the Expression Language function? In my opinion, that's the real "problem" here, as it's a static method. I would *think* you could run into an issue if you used the lazy initiator and had an instance of the expression language form before the tag form.

rombert committed:
SLING-8266 - Stop embedding ESAPI

Stop exposing instance methods from XSSSupport, they don't make sense.

jsedding commented on src/main/java/org/apache/sling/scripting/jsp/taglib/helpers/XSSSupport.java:
```diff
@@ -16,56 +16,89 @@
+    @Reference
+    private XSSAPI xssApi;
+    
+    private static XSSAPI XSS_API;
```
I believe this field needs to be volatile.

rombert replied:
Good catch, thank you.

rombert committed:
SLING-8266 - Stop embedding ESAPI

Adapt tests by porting some of them to Sling Mocks. With the new
requirements added by the XSS API bundle, we need to pull in a more
recent version of the api bundle. With that, the Sling Testing Tools
mocks no longer provided all the needed methods so just switch to Sling
mocks outright.

rombert committed:
SLING-8266 - Stop embedding ESAPI

Make XSS_API field volatile.

rombert commented:
@kwin - the component is implicitly immediate, see https://osgi.org/specification/osgi.cmpn/7.0.0/service.component.html#d0e36881 (emphasis added)

> A component is an immediate component **if it is not a factory component and either does not specify a service** or specifies a service and the immediate attribute of the component element set to true.

rombert commented:
After adapting the tests I see that two checks are failing

```
Failed tests: 
  TestEncodeTag.testEncode:185 expected:<&amp;amp[&#x3b;Hello World&#x21;]> but was:<&amp;amp[;Hello World!]>
  TestSlingFunctions.testEncode:90 expected:<&amp;nbsp[&#x3b;Here is some text&#x21;]> but was:<&amp;nbsp[;Here is some text!]>
```

It looks like the encoding functions were more eager in the taglib setup but are more relaxed in the XSS API bundle ( e.g. don't encode `!` ). I would suggest to adapt the tests to use the form returned by the XSS API as it brings the encoding close to Sling behaviour in general.

rombert commented:
@kwin - the EL function that @klcodanr was referring to is https://github.com/apache/sling-org-apache-sling-scripting-jsp-taglib/blob/060ce101e33ddf03feb450c3ccf4f6e3ec423f32/src/main/java/org/apache/sling/scripting/jsp/taglib/SlingFunctions.java#L84-L86 . In here we only have two string arguments, so nothing extra can be passed. We have no implicit objects in EL functions, and adding a new parameter is a breaking change.

This is why the implementation still uses the static `XSSSupport` methods and why I set up that specific class as an OSGi component.

rombert committed:
SLING-8266 - Stop embedding ESAPI

Adjust tests to XSS API encoding configuration and add license heder to
XSSSupportRule.

rombert committed:
SLING-8266 - Stop embedding ESAPI

Fix compilation error

rombert committed:
SLING-8266 - Stop embedding ESAPI

Fix compilation error (v2)

rombert commented:
A SonarQube report for the changes added _only by this pull request_ was generated. Please review it at https://builds.apache.org/job/Sling/job/sling-org-apache-sling-scripting-jsp-taglib/job/PR-1/5/artifact/target/sonar/issues-report/issues-report-light.html

rombert committed:
SLING-8266 - Stop embedding ESAPI

Remove unused imports.

rombert committed:
SLING-8019 - Use XSS API

Stop exporting the XSSSupport component, this will be contributed by a
compat bundle.

rombert commented:
Following @cziegeler 's suggestion I've moved 'helper' package classes to 'internal' and will create a new compat bundle for the old `XSSSupport` class.

The reason (from [SLING-8019](https://issues.apache.org/jira/browse/SLING-8019) ) is that there is no need for a second XSS API now that we have the proper XSS bundle.

rombert commented:
A SonarQube report for the changes added _only by this pull request_ was generated. Please review it at https://builds.apache.org/job/Sling/job/sling-org-apache-sling-scripting-jsp-taglib/job/PR-1/7/artifact/target/sonar/issues-report/issues-report-light.html

rombert committed:
SLING-8266 - Stop embedding ESAPI

Remove private constructor from XSSSupport.

rombert commented:
A SonarQube report for the changes added _only by this pull request_ was generated. Please review it at https://builds.apache.org/job/Sling/job/sling-org-apache-sling-scripting-jsp-taglib/job/PR-1/8/artifact/target/sonar/issues-report/issues-report-light.html

rombert commented:
A SonarQube report for the changes added _only by this pull request_ was generated. Please review it at https://builds.apache.org/job/Sling/job/sling-org-apache-sling-scripting-jsp-taglib/job/PR-1/9/artifact/target/sonar/issues-report/issues-report-light.html

rombert commented:
Should be rebased on top of #2 to make sure we don't introduce BC breaking changes in the next release.

rombert commented:
A SonarQube report for the changes added _only by this pull request_ was generated. Please review it at https://builds.apache.org/job/Sling/job/sling-org-apache-sling-scripting-jsp-taglib/job/PR-1/10/artifact/target/sonar/issues-report/issues-report-light.html

rombert committed:
Merge remote-tracking branch 'origin/master' into feature/SLING-8266

rombert commented:
Will rebase + squash this locally, thanks all for the reviews.
